astropy/io/fits/file.py

```diff
@@ -426,13 +430,33 @@ def _open_fileobj(self, fileobj, mode, overwrite):
             self._file = fileobj
         elif isfile(fileobj):
             self._file = fileobj_open(self.name, IO_FITS_MODES[mode])
-        else:
-            self._file = gzip.open(self.name, IO_FITS_MODES[mode])
 
-        if fmode == 'ab+':
-            # Return to the beginning of the file--in Python 3 when opening in
-            # append mode the file pointer is at the end of the file
+        # Attempt to determine if the file represented by the open file object
+        # is compressed
+        try:
+            magic = self._file.read(4)
             self._file.seek(0)
+        except (IOError,OSError):
+            return
+
+        if magic.startswith(GZIP_MAGIC):
```

Comment:
It's not identical but I'll see if there's a way to share more code between the two cases.